File: scripts/feet_to_odom.py
#!/bin/env python3

import rclpy
from rclpy.node import Node

from go2_odometry.msg import OdometryVector
from unitree_go.msg import LowState
from geometry_msgs.msg import PoseWithCovariance
from scipy.spatial.transform import Rotation

import numpy as np
import pinocchio as pin
from go2_description import loadGo2

class FeetToOdom(Node):

    def __init__(self):
        super().__init__('feet_to_odom')

        self.pos_publisher = self.create_publisher(OdometryVector, 'odometry/feet_pos', 10)
        self.subscription = self.create_subscription(
            LowState,
            '/lowstate',
            self.listener_callback,
            10)

        self.robot = loadGo2()
        self.foot_frame_name = [prefix + "_foot" for prefix in ["FL", "FR", "RL", "RR"]]
        self.foot_frame_id = [self.robot.model.getFrameId(frame_name) for frame_name in self.foot_frame_name]
        self.imu_frame_id = self.robot.model.getFrameId("imu")

    def _unitree_to_urdf_vec(self, vec):
        return  [vec[3],  vec[4],  vec[5],
                 vec[0],  vec[1],  vec[2],
                 vec[9],  vec[10], vec[11],
                 vec[6],  vec[7],  vec[8],]


    def listener_callback(self, state_msg):
        # Get sensor measurement
        q_unitree = [j.q for j in state_msg.motor_state[:12]]
        v_unitree = [j.dq for j in state_msg.motor_state[:12]]
        fc_unitree = state_msg.foot_force

        # Rearrange joints according to urdf
        q = np.array([0]*6 + [1] + self._unitree_to_urdf_vec(q_unitree))
        v = np.array([0]*6 + self._unitree_to_urdf_vec(v_unitree))
        f_contact = [fc_unitree[i] for i in [1, 0, 3, 2]]

        # Compute positions and velocities
        ## f = foot, i = imu, b = base
        self.robot.forwardKinematics(q, v)
        pin.updateFramePlacements(self.robot.model, self.robot.data)
        bMf_list = [self.robot.data.oMf[id] for id in self.foot_frame_id]

        # Make message
        pos_msg = OdometryVector()
        pos_msg.header.stamp = self.get_clock().now().to_msg()
        pos_list = []
        feet_list = []
        for i in range(4):
            if(f_contact[i]<10):
                continue # Feet in the air : skip

            feet_list.append(self.foot_frame_name[i])
            pose_foot = PoseWithCovariance()
            pose_foot.covariance = [0.] * 36

            pose_foot.pose.position.x = bMf_list[i].translation[0]
            pose_foot.pose.position.y = bMf_list[i].translation[1]
            pose_foot.pose.position.z = bMf_list[i].translation[2]

            foot_rotation = bMf_list[i].rotation
            r = Rotation.from_matrix([[foot_rotation[0][0], foot_rotation[0][1], foot_rotation[0][2]],
                                  [foot_rotation[1][0], foot_rotation[1][1], foot_rotation[1][2]],
                                  [foot_rotation[2][0], foot_rotation[2][1], foot_rotation[2][2]]])
            
            qx, qy, qz, qw = r.as_quat()
            pose_foot.pose.orientation.x = qx
            pose_foot.pose.orientation.y = qy
            pose_foot.pose.orientation.z = qz
            pose_foot.pose.orientation.w = qw

            pos_list.append(pose_foot)

        pos_msg.feet_names = feet_list
        pos_msg.pose_vec = pos_list

        self.pos_publisher.publish(pos_msg)

def main(args=None):
    rclpy.init(args=args)

    feet_to_odom = FeetToOdom()

    rclpy.spin(feet_to_odom)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    feet_to_odom.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/feet_to_odom.cpp
#include <rclcpp/rclcpp.hpp>
#include <rmw/qos_profiles.h>
#include <rclcpp/qos.hpp>

#include <unitree_go/msg/low_state.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_with_covariance.hpp>
#include <pinocchio/fwd.hpp>
#include <pinocchio/algorithm/frames.hpp>
#include <pinocchio/algorithm/model.hpp>
#include <pinocchio/algorithm/compute-all-terms.hpp>
#include <pinocchio/parsers/urdf.hpp>

#include "go2_odometry/msg/odometry_vector.hpp"

class FeetToOdom : public rclcpp::Node
{
public:
    explicit FeetToOdom()
    : Node("inekf")
    {
        rmw_qos_profile_t qos = rmw_qos_profile_sensor_data;
        qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
        qos.history = RMW_QOS_POLICY_HISTORY_KEEP_LAST;
        
        // Write subscribers to feet odometry and IMU
        pos_feet_publisher_ = this->create_publisher<go2_odometry::msg::OdometryVector>(
            "odometry/feet_pos", 
            rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(qos), qos)
        );
        
        lowstate_subscriber_ = this->create_subscription<unitree_go::msg::LowState>(
            "lowstate", 
            10, 
            std::bind(&FeetToOdom::listener_callback, this, std::placeholders::_1)
        );

        unitree_to_urdf_vec_ = {3, 4, 5,
            0, 1, 2,
            9, 10, 11,
            6, 7, 8};
        
        force_order_ = {1, 0, 3, 2};

        q_.resize(19);
        q_.setZero();
        q_[6] = 1;
        f_contact_.resize(4);
        f_contact_.setZero();

        // Load robot
        std::string urdf_path = GO2_DESCRIPTION_MODEL_DIR "/go2.urdf";
        std::string srdf_path = GO2_DESCRIPTION_MODEL_DIR "/go2.srdf";
        std::string base_joint_name ="root_joint";
        
        pinocchio::urdf::buildModel(urdf_path, pinocchio::JointModelFreeFlyer(), model_);
        data_ = pinocchio::Data(model_);
        
        foot_frame_names_ = {"FL_foot", "FR_foot", "RL_foot", "RR_foot"};
        for (auto it = foot_frame_names_.cbegin(); it != foot_frame_names_.cend(); ++it) {
            foot_frame_ids_.push_back(model_.getFrameId(*it));
        }
    }

    void listener_callback(const unitree_go::msg::LowState::SharedPtr msg)
    {   
        // Rearrange joints according to urdf
        for (std::size_t i = 0; i < 12; i ++) {
            q_[7 + i] = msg->motor_state[unitree_to_urdf_vec_[i]].q;
        }
        for (std::size_t i = 0; i < 4; i ++) {
            f_contact_[i] = msg->foot_force[force_order_[i]];
        }

        // Compute positions and velocities
        // f = foot, i = imu, b = base
        pinocchio::forwardKinematics(model_, data_, q_);
        pinocchio::updateFramePlacements(model_, data_);
        std::vector<pinocchio::SE3> bMf_list;
        for (auto it = foot_frame_ids_.cbegin(); it != foot_frame_ids_.cend(); ++it) {
            bMf_list.push_back(data_.oMf[*it]);
        }
        
        // Create message
        go2_odometry::msg::OdometryVector pose_msg;
        pose_msg.header.stamp = this->get_clock()->now();

        std::vector<geometry_msgs::msg::PoseWithCovariance> pose_list;
        std::vector<bool> contact_states;
        for (std::size_t i = 0; i < 4; i++) {
            if (f_contact_[i] > 20) contact_states.push_back(true);
            else contact_states.push_back(false);

            geometry_msgs::msg::PoseWithCovariance pose_foot;
            pose_foot.covariance.fill(0);

            pose_foot.pose.position.x = bMf_list[i].translation()[0];
            pose_foot.pose.position.y = bMf_list[i].translation()[1];
            pose_foot.pose.position.z = bMf_list[i].translation()[2];

            Eigen::Quaternion<double> quat(bMf_list[i].rotation());
            
            pose_foot.pose.orientation.x = quat.x();
            pose_foot.pose.orientation.y = quat.y();
            pose_foot.pose.orientation.z = quat.z();
            pose_foot.pose.orientation.w = quat.w();

            pose_list.push_back(pose_foot);
        }
        pose_msg.contact_states = contact_states;
        pose_msg.pose_vec = pose_list;

        pos_feet_publisher_->publish(pose_msg);
    }

    rclcpp::Publisher<go2_odometry::msg::OdometryVector>::SharedPtr pos_feet_publisher_;
    rclcpp::Subscription<unitree_go::msg::LowState>::SharedPtr lowstate_subscriber_;

    std::vector<long> unitree_to_urdf_vec_;
    std::vector<long > force_order_;
    Eigen::VectorXd q_;
    Eigen::VectorXd f_contact_;
    std::vector<std::size_t> foot_frame_ids_;
    std::vector<std::string> foot_frame_names_;
    pinocchio::Model model_;
    pinocchio::Data data_;
};


int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FeetToOdom>());
  rclcpp::shutdown();
  return 0;
}